medusa/databases/failed_db.py

```diff
@@ -111,3 +112,87 @@ def execute(self):
+
+        self.translate_status()
+        self.inc_major_version()
+
+    def translate_status(self):
+        """
+        Add columns status and quality.
+
+        Translate composite status/quality from old_status to the new fields.
+        """
+        log.info(u'Transforming old_status (composite) to separated fields status + quality.')
+        self.connection.action('DROP TABLE IF EXISTS new_history;')
+
+        self.connection.action('CREATE TABLE new_history (date NUMERIC, size NUMERIC, release TEXT, provider TEXT, '
+                               'status NUMERIC DEFAULT -1, quality NUMERIC DEFAULT 0, showid NUMERIC DEFAULT -1, '
+                               'season NUMERIC DEFAULT -1, episode NUMERIC DEFAULT -1, indexer_id NUMERIC)')
+
+        self.connection.action('INSERT INTO new_history (date, size, release, provider, '
+                               'status, quality, showid, season, episode, indexer_id) '
+                               'SELECT date, size, release, provider, '
+                               'old_status, -1, showid, season, episode, indexer_id '
+                               'FROM history;')
+        self.connection.action('DROP TABLE IF EXISTS history;')
+        self.connection.action('ALTER TABLE new_history RENAME TO history;')
+        self.connection.action('DROP TABLE IF EXISTS new_history;')
+
+        sql_results = self.connection.select('SELECT status FROM history GROUP BY status;')
+        for result in sql_results:
+            split = Quality.split_composite_status(result[b'status'])
+            self.connection.action('UPDATE history SET status = ?, quality = ? WHERE status = ?;',
+                                   [split.status, split.quality, result[b'status']])
+
+    def inc_major_version(self):
+        major_version, minor_version = self.connection.version
+        major_version += 1
+        self.connection.action("UPDATE db_version SET db_version = ?;", [major_version])
+        return self.connection.version
+
+
+class ShiftQualities(UpdateHistoryTableQuality):
+    """Shift all qualities one place to the left."""
+
+    def test(self):
+        """Test if the version is at least 3."""
+        return self.connection.version >= (3, None)
+
+    def execute(self):
+        utils.backup_database(self.connection.path, self.connection.version)
+
+        self.shift_history_qualities()
+        self.update_status_unknown()
+        self.inc_major_version()
+
+    def shift_history_qualities(self):
+        """
+        Shift all qualities << 1.
+
+        This makes it possible to set UNKNOWN as 1, making it the lowest quality.
+        """
+        log.info('Shift qualities in history one place to the left.')
+        sql_results = self.connection.select("SELECT quality FROM history GROUP BY quality ORDER BY quality DESC;")
+        for result in sql_results:
+            quality = result[b'quality']
+            new_quality = quality << 1
```

Comment:
I just woke up, but... You're supposed to shift right by one bit, not left. Right? :thinking: 
This applies to other locations too.

Reply:
No, how do you come to that conclusion now?